include/pisa/topk_queue.hpp

```diff
@@ -172,6 +181,26 @@ struct topk_queue {
         }
     }
 
+    // We use our own function (as opposed to `std::heap_push`), to ensure that
+    // heap implementation is consistent across standard libraries.
```

Comment:
What does `to ensure that heap implementation is consistent across standard libraries` mean?

Reply:
Means that if I compile on libstdc++ 12 and you compile on libc++ 10, then both of us have the same logic, and thus if you run the same exact algorithm on the same exact collection, etc., you will get the same results.

With `std::push`, the implementation may change from version to version and implementation to implementation (LLVM/GNU/MS/Apple) so it's not necessarily reproducible across environments.

Using pair ordering (score, doc) woulds of course solve all of that, but that's not efficient.

Reply:
Understand! Makes sense. Right now we are not aware of differences in the implementation  of libc++ 10 and libstdc++ 12, right?

My major objection to this approach is that if we stick to this paradigm, we would need to reimplement the entire STL.

Reply:
> My major objection to this approach is that if we stick to this paradigm, we would need to reimplement the entire STL

Normally, I would agree, but this is somewhat special circumstances. Technically speaking, we probably would want to strictly define the order (score + docid) but we're not because of efficiency. But also we are already implementing the more complex pop operation, so we might as well finish the job. It's very little effort for possibly a significant improvement for someone down the line.

Reply:
Let's put it this way: you're doing research, and have some results, and then you recompile after system upgrade or something, and it doesn't match; it's not life changing, but that's very annoying imo, and I personally see no reason not to fix that with one little function. It could save someone a headache down the line.